src/Conversion/ONNXToKrnl/Tensor/ScatterND.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===--------------- ScatterND.cpp - Lowering ScatterND Op ----------------===//
+//
+// Copyright 2022 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX ScatterND Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+#include "src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+struct ONNXScatterNDOpLowering : public ConversionPattern {
+  ONNXScatterNDOpLowering(TypeConverter &typeConverter, MLIRContext *ctx)
+      : ConversionPattern(
+            typeConverter, ONNXScatterNDOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXScatterNDOpAdaptor operandAdaptor(operands);
+    Location loc = op->getLoc();
+
+    // Operands and attributes.
+    Value data = operandAdaptor.data();
+    Value updates = operandAdaptor.updates();
+    Value indices = operandAdaptor.indices();
+    auto dataType = data.getType().cast<ShapedType>();
+    auto indicesType = indices.getType().cast<ShapedType>();
+    auto updatesType = updates.getType().cast<ShapedType>();
+    int64_t dataRank = dataType.getRank();
+    int64_t updatesRank = updatesType.getRank();
+    int64_t indicesRank = indicesType.getRank();
+
+    assert(dataRank >= 1 && "The rank of 'data' must be >= 1");
+    assert(indicesRank >= 1 && "The rank of 'indices' must be >= 1");
```

Comment:
I assume that these input dependent assert where tested in the verifier and are here only for redundancy (which is perfectly fine IMO).
Code below seems to confirm that the verifiers indeed tested these conditions.

Reply:
Yes that is correct.  The idea is to diagnose error conditions in the verifier, and in shape inference defensively assert that the "code" is as expected.